dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ExpressionEncoder.java

```diff
@@ -0,0 +1,444 @@
+            BitvectorFormulaManager bvmgr = fmgr.getBitvectorFormulaManager();
+            BitvectorFormula bv1 = (BitvectorFormula) lhs;
+            BitvectorFormula bv2 = (BitvectorFormula) rhs;
+            switch (op) {
+                case EQ:
+                    return bvmgr.equal(bv1, bv2);
+                case NEQ:
+                    return bmgr.not(bvmgr.equal(bv1, bv2));
+                case LT:
+                case ULT:
+                    return bvmgr.lessThan(bv1, bv2, op.equals(COpBin.LT));
+                case LTE:
+                case ULTE:
+                    return bvmgr.lessOrEquals(bv1, bv2, op.equals(COpBin.LTE));
+                case GT:
+                case UGT:
+                    return bvmgr.greaterThan(bv1, bv2, op.equals(COpBin.GT));
+                case GTE:
+                case UGTE:
+                    return bvmgr.greaterOrEquals(bv1, bv2, op.equals(COpBin.GTE));
+            }
+        }
+        throw new UnsupportedOperationException("Encoding not supported for COpBin: " + lhs + " " + op + " " + rhs);
+    }
+
+    @Override
+    public Formula visit(Atom atom) {
+        Formula lhs = encodeInteger(atom.getLHS());
+        Formula rhs = encodeInteger(atom.getRHS());
+        return encodeAtom(atom.getOp(), lhs, rhs, formulaManager);
+    }
+
+    @Override
+    public Formula visit(BConst bConst) {
+        return booleanFormulaManager.makeBoolean(bConst.getValue());
+    }
+
+    @Override
+    public Formula visit(BExprBin bBin) {
+        BooleanFormula lhs = encodeBoolean(bBin.getLHS());
+        BooleanFormula rhs = encodeBoolean(bBin.getRHS());
+        switch (bBin.getOp()) {
+            case AND:
+                return booleanFormulaManager.and(lhs, rhs);
+            case OR:
+                return booleanFormulaManager.or(lhs, rhs);
+        }
+        throw new UnsupportedOperationException("Encoding of not supported for BOpBin " + bBin.getOp());
+    }
+
+    @Override
+    public Formula visit(BExprUn bUn) {
+        BooleanFormula inner = encodeBoolean(bUn.getInner());
+        return booleanFormulaManager.not(inner);
+    }
+
+    @Override
+    public Formula visit(BNonDet bNonDet) {
+        return booleanFormulaManager.makeVariable(Integer.toString(bNonDet.hashCode()));
+    }
+
+    @Override
+    public Formula visit(IValue iValue) {
+        return iValue.getPrecision() > 0
+                ? bitvectorFormulaManager.makeBitvector(iValue.getPrecision(), iValue.getValue())
+                : integerFormulaManager.makeNumber(iValue.getValue());
+    }
+
+    @Override
+    public Formula visit(IExprBin iBin) {
+        Formula lhs = encodeInteger(iBin.getLHS());
+        Formula rhs = encodeInteger(iBin.getRHS());
+        // Some SMT solvers do not support certain theories.
+        // Calling the constructor of the manager in such solvers results in an Exception.
+        // Thus we initialize the manager inside the branches
```

Comment:
These lines seem to be outdated and make me wonder if the initializer in `ExpressionEncoder` would work for those cases there the solver does not support the theory

Reply:
It would not work properly.  `getIntegerFormulaManager()` will throw an `UnsupportedOperation`.  I am going to fix this.